[SPARK-33212][FOLLOW-UP][BUILD] Bring back duplicate dependency check and add more strict Hadoop version check

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

1. Add back Maven enforcer for duplicate dependencies check
2. More strict check on Hadoop versions which support shaded client in `IsolatedClientLoader`. To do proper version check, this adds a util function `majorMinorPatchVersion` to extract major/minor/patch version from a string.
3. Cleanup unnecessary code

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

The Maven enforcer was removed as part of #30556. This proposes to add it back. 

Also, Hadoop shaded client doesn't work in certain cases (see [these comments](https://github.com/apache/spark/pull/30701#discussion_r558522227) for details). This strictly checks that the current Hadoop version (i.e., 3.2.2 at the moment) has good support of shaded client or otherwise fallback to old unshaded ones.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Existing tests.

sunchao committed:
SPARK-33212 follow-up

xkrogen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -107,12 +107,17 @@ private[hive] object IsolatedClientLoader extends Logging {
   }
 
+  def supportHadoopShadedClient(hadoopVersion: String): Boolean = hadoopVersion match {
+    case "3.2.2" => true
```
Maintaining a list of versions is probably going to become a pain, is there a semantic version comparison function hanging around somewhere we can use to verify `isMoreRecentVersion(hadoopVersion, "3.2.2")` ?

sunchao replied:
Yup that is a good idea. I'll update it later.

xkrogen reviewed (changes requested):
Thanks for the fast follow-on Chao!

SparkQA commented:
**[Test build #134130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134130/testReport)** for PR 31203 at commit [`062b7bc`](https://github.com/apache/spark/commit/062b7bc2974ddb0ab14681b155cb8d10b52af640).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38713/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38713/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38713/

SparkQA commented:
**[Test build #134130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134130/testReport)** for PR 31203 at commit [`062b7bc`](https://github.com/apache/spark/commit/062b7bc2974ddb0ab14681b155cb8d10b52af640).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134130/

SparkQA commented:
**[Test build #134146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134146/testReport)** for PR 31203 at commit [`b5d82b7`](https://github.com/apache/spark/commit/b5d82b7ab02248a4bc22df29d1deaf9c84abe184).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38729/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38729/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38729/

SparkQA commented:
**[Test build #134146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134146/testReport)** for PR 31203 at commit [`b5d82b7`](https://github.com/apache/spark/commit/b5d82b7ab02248a4bc22df29d1deaf9c84abe184).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134146/

sunchao committed:
add version parsing

sunchao commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -88,31 +89,71 @@ private[hive] object IsolatedClientLoader extends Logging {
-    case "3.0" | "3.0.0" => hive.v3_0
-    case "3.1" | "3.1.0" | "3.1.1" | "3.1.2" => hive.v3_1
-    case version =>
+  def hiveVersion(version: String): HiveVersion = {
```
This is probably out of the scope of this PR. I'll open a new one if we agree this is the right thing to do.

xkrogen replied:
Seems a good improvement to me, that table is pretty unsightly as-is.

SparkQA commented:
**[Test build #134382 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134382/testReport)** for PR 31203 at commit [`6657c87`](https://github.com/apache/spark/commit/6657c874d48c16513b8cef4a0b8ee9df6b2c531d).

xkrogen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -88,31 +89,71 @@ private[hive] object IsolatedClientLoader extends Logging {
+  def supportHadoopShadedClient(hadoopVersion: String): Boolean = {
+    getVersionParts(hadoopVersion).exists {
+      case (3, 2, v) if v >= 2 => true
+      case _ => false
```
maybe
```
case (maj, _, _) if maj > 3 => true
case (3, min, _) if min > 2 => true
case (3, 2, patch) if patch >=2 => true
```
Seems like we can reasonably assume that future versions of Hadoop will support the shaded client?

sunchao replied:
I think we'd better to wait until the future versions come out before changing this (so that we can verify firs). For instance, Hadoop 3.3.0 currently doesn't support shaded client (due to the hadoop-aws issue). But yeah the Hadoop 3.2.2+ should support the shaded client assuming there's no regression.

xkrogen replied:
Interesting... I just worry about forgetting to update this if/when we bump the Hadoop version in the future and causing a regression. Has the `hadoop-aws` fix made it into Hadoop 3.3.1? If so, can we reasonably assume that 3.2.2+, 3.3.1+, and 3.4.0+ will have it?

sunchao replied:
I think your concern is valid. One thing we can do is perhaps adding a test to make sure that the built-in Hadoop version is always compatible with the shaded client. So that in future if we upgrade Hadoop version & forget to do this, the test will break.

sunchao replied:
And yes we can assume that 3.2.2+, 3.3.1+ and 3.4.0+ will all have the fix.

xkrogen replied:
Excellent idea on adding a compatibility test for the built-in Hadoop version!

xkrogen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -88,31 +89,71 @@ private[hive] object IsolatedClientLoader extends Logging {
+    val matcher = VERSION_PATTERN.matcher(version)
+    if (matcher.matches() && matcher.groupCount() == 3) {
+      val major = matcher.group(1).toInt
+      val minor = if (matcher.group(2) == null) 0 else matcher.group(2).toInt
```
Maybe:
```
val minor = Option(matcher.group(2)).map(_.toInt).getOrElse(0)
```

sunchao replied:
I like this. Will change.

xkrogen reviewed (changes requested):
I wonder if this should go into `VersionUtils`? That does claim to be specifically for working with _Spark_ version strings, but it seems relevant...

sunchao commented:
> I wonder if this should go into VersionUtils? That does claim to be specifically for working with Spark version strings, but it seems relevant...

Oops I wasn't even aware of `VersionUtils`. Yes I agree it seems a better place to put this code. Let me update the PR later.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134381/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38968/

SparkQA commented:
**[Test build #134382 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134382/testReport)** for PR 31203 at commit [`6657c87`](https://github.com/apache/spark/commit/6657c874d48c16513b8cef4a0b8ee9df6b2c531d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134382/

sunchao committed:
address comments

- move version parsing to `VersionUtils`
- add compatibility test
- use `Option` for null

SparkQA commented:
**[Test build #134405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134405/testReport)** for PR 31203 at commit [`9b8a911`](https://github.com/apache/spark/commit/9b8a911dcf1aa95ca39ca7e64648c39573f5451b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38991/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38991/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38991/

SparkQA commented:
**[Test build #134405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134405/testReport)** for PR 31203 at commit [`9b8a911`](https://github.com/apache/spark/commit/9b8a911dcf1aa95ca39ca7e64648c39573f5451b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134405/

xkrogen commented on core/src/test/scala/org/apache/spark/util/VersionUtilsSuite.scala:
```diff
@@ -98,4 +98,21 @@ class VersionUtilsSuite extends SparkFunSuite {
+    assert(VersionUtils.majorMinorPatchVersion("3").contains((3, 0, 0)))
+
+    // illegal cases
+    assert(VersionUtils.majorMinorPatchVersion("ABC").isEmpty)
+    assert(VersionUtils.majorMinorPatchVersion("3X").isEmpty)
+    assert(VersionUtils.majorMinorPatchVersion("3.2-SNAPSHOT").isEmpty)
+    assert(VersionUtils.majorMinorPatchVersion("3.2ABC").isEmpty)
+    assert(VersionUtils.majorMinorPatchVersion("3-ABC").isEmpty)
+    assert(VersionUtils.majorMinorPatchVersion("3.2.4XYZ").isEmpty)
```
You can consider simplifying this a little like:
```
Seq("ABC", "3X", ...).foreach { versionStr =>
  assert(VersionUtils.majorMinorPatchVersion(versionStr).isEmpty, s"version $versionStr")
}
```

xkrogen commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HadoopVersionInfoSuite.scala:
```diff
@@ -68,4 +69,22 @@ class HadoopVersionInfoSuite extends SparkFunSuite {
   }
+
+  test("SPARK-32212: test supportHadoopShadedClient()") {
+    assert(IsolatedClientLoader.supportHadoopShadedClient("3.2.2"))
+    assert(IsolatedClientLoader.supportHadoopShadedClient("3.2.3"))
+    assert(IsolatedClientLoader.supportHadoopShadedClient("3.2.2.1"))
+    assert(IsolatedClientLoader.supportHadoopShadedClient("3.2.2-XYZ"))
+    assert(IsolatedClientLoader.supportHadoopShadedClient("3.2.2.4-SNAPSHOT"))
```
Maybe consider shortening like:
```
Seq("3.2.2", "3.2.3", ...).foreach { versionStr =>
  assert(IsolatedClientLoader.supportHadoopShadedClient(versionStr), s"version $versionStr")
}
```
(and same for the negative cases)

xkrogen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -88,31 +88,40 @@ private[hive] object IsolatedClientLoader extends Logging {
+    }
+  }
+
+  def supportHadoopShadedClient(hadoopVersion: String): Boolean = {
```
Should this be `supportsHadoopShadedClient` instead of `supportHadoopShadedClient` ?

xkrogen commented on core/src/main/scala/org/apache/spark/util/VersionUtils.scala:
```diff
@@ -63,4 +64,37 @@ private[spark] object VersionUtils {
+
+  /**
+   * Retrieves the major, minor and patch parts from the input `version`. Returns `None` if the
+   * input is not of a valid format.
```
I think we should mention that minor/patch versions are filled in as 0 if they're not found. This is different from the behavior of other methods in this class (e.g. majorMinor will give an error if minor is not present)

sunchao replied:
Yeah good point.

xkrogen commented on core/src/main/scala/org/apache/spark/util/VersionUtils.scala:
```diff
@@ -63,4 +64,37 @@ private[spark] object VersionUtils {
+   *          input is a valid version. `None` otherwise.
+   */
+  def majorMinorPatchVersion(version: String): Option[(Int, Int, Int)] = {
+    majorMinorPatchRegex.findFirstMatchIn(version) match {
+      case Some(m) =>
+        val major = m.group(1).toInt
+        val minor = Option(m.group(2)).map(_.toInt).getOrElse(0)
+        val patch = Option(m.group(3)).map(_.toInt).getOrElse(0)
+        Some((major, minor, patch))
+      case None => None
```
Maybe a little cleaner to use `map`, the `None => None` is a little awkward:
```
    majorMinorPatchRegex.findFirstMatchIn(version).map { m =>
      val major = m.group(1).toInt
      val minor = Option(m.group(2)).map(_.toInt).getOrElse(0)
      val patch = Option(m.group(3)).map(_.toInt).getOrElse(0)
      (major, minor, patch)
    }
```

sunchao committed:
address comments & remove hive changes

SparkQA commented:
**[Test build #134499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134499/testReport)** for PR 31203 at commit [`396130b`](https://github.com/apache/spark/commit/396130b5c6fe9a671e1a0ca90b533f5a4d6724de).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39085/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39085/

SparkQA commented:
**[Test build #134499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134499/testReport)** for PR 31203 at commit [`396130b`](https://github.com/apache/spark/commit/396130b5c6fe9a671e1a0ca90b533f5a4d6724de).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39085/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134499/

xkrogen commented:
Changes LGTM thanks @sunchao !

sunchao commented:
Thanks @xkrogen for the thorough review! @dongjoon-hyun @viirya could you take a look?

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/util/VersionUtils.scala:
```diff
@@ -63,4 +64,36 @@ private[spark] object VersionUtils {
+   *  - 2.4   (extracts to (2, 4, 0))
+   *  - 3.2.2   (extracts to (3, 2, 2))
+   *  - 3.2.2.4   (extracts to 3, 2, 2))
+   *  - 3.3.1-SNAPSHOT   (extracts to (3, 2, 2))
```
```scala
-   *  - 3.3.1-SNAPSHOT   (extracts to (3, 2, 2))
+   *  - 3.2.1-SNAPSHOT   (extracts to (3, 2, 2))
```

viirya replied:
```
*  - 3.2.1-SNAPSHOT   (extracts to (3, 2, 1))?
```

sunchao replied:
oops - nice catch!

sunchao replied:
I updated it to `*  - 3.3.1-SNAPSHOT   (extracts to (3, 3, 1))`

dongjoon-hyun commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -107,12 +107,19 @@ private[hive] object IsolatedClientLoader extends Logging {
+  def supportsHadoopShadedClient(hadoopVersion: String): Boolean = {
+    VersionUtils.majorMinorPatchVersion(hadoopVersion).exists {
+      case (3, 2, v) if v >= 2 => true
+      case _ => false
```
Got it. I agree that we need to change this again.
> And yes we can assume that 3.2.2+, 3.3.1+ and 3.4.0+ will all have the fix.

dongjoon-hyun reviewed (approved):
+1, LGTM (with only one minor comment about comment typo)

sunchao committed:
fix comment

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -123,22 +130,14 @@ private[hive] object IsolatedClientLoader extends Logging {
         .map(a => s"org.apache.hive:$a:${version.fullVersion}") ++
       Seq("com.google.guava:guava:14.0.1") ++ hadoopJarNames
 
-    val extraExclusions = if (hadoopVersion.startsWith("3")) {
-      // this introduced from lower version of Hive could conflict with jars in Hadoop 3.2+, so
-      // exclude here in favor of the ones in Hadoop 3.2+
-      Seq("org.apache.hadoop:hadoop-auth")
-    } else {
```
No need to exclude hadoop-auth anymore?

sunchao replied:
Yea it's no longer needed. Please see [here](https://github.com/apache/spark/pull/30701#discussion_r558529521) and [here](https://github.com/apache/spark/pull/29843/files#r509699879) for the context.

viirya reviewed (approved):
looks good with one question.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134524/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39109/

dongjoon-hyun commented:
Since the last question is addressed, I'll merge this. Thanks!
